File: breadth-first-search/average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.go
package averageoflevelsinbinarytree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func averageOfLevels(root *TreeNode) []float64 {
	avg := []float64{}

	queue := []*TreeNode{}

	queue = append(queue, root)

	for len(queue) > 0 {
		n := len(queue)
		count := n
		sum := 0

		for count > 0 {
			curr := queue[0]
			queue = queue[1:]

			sum += curr.Val

			if curr.Left != nil {
				queue = append(queue, curr.Left)
			}

			if curr.Right != nil {
				queue = append(queue, curr.Right)
			}
			count--
		}
		avg = append(avg, float64(sum)/float64(n))
	}
	return avg
}

File: breadth-first-search/average-of-levels-in-binary-tree/average-of-levels-in-binary-tree.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
  public List<Double> averageOfLevels(TreeNode root) {
    List<Double> avg = new ArrayList<>();

    Queue<TreeNode> queue = new ArrayDeque<>();

    queue.add(root);

    while (!queue.isEmpty()) {
      int n = queue.size();
      int count = n;
      long sum = 0;

      while (count > 0) {
        TreeNode curr = queue.poll();

        sum += curr.val;

        if (curr.left != null) {
          queue.add(curr.left);
        }

        if (curr.right != null) {
          queue.add(curr.right);
        }
        count--;
      }
      avg.add((double) sum / n);
    }

    return avg;
  }
}
